overhaul contributing based on feedback from evan (#393)
OpsLevel/cli: CONTRIBUTING.md

# Contributing to opslevel-cli

👋 Welcome, and thank you for your interest in contributing to `opslevel-cli`! This guide will help you ramp up, propose changes, develop locally, and contribute code effectively.

---

## Table of Contents

1. [About the CLI](#about-the-cli)
2. [Getting Started](#getting-started)
3. [Development Workflow](#development-workflow)
4. [Proposing and Submitting Changes](#proposing-and-submitting-changes)
5. [Command Architecture & Style](#command-architecture--style)
6. [Testing & Tooling](#testing--tooling)
7. [Release Process](#release-process)

---

## About the CLI

The `opslevel-cli` is a command-line interface for interacting with the [OpsLevel](https://www.opslevel.com) API. It helps engineers automate, inspect, and manage their service catalog, ownership, checks, and more.

### Architecture

> - **[Cobra](https://github.com/spf13/cobra)** powers our CLI framework. Commands are defined as Go structs with handlers, descriptions, and flags.

Author: So the public docs for Cobra state

```
Cobra is a library for creating powerful modern CLI applications.

Cobra is used in many Go projects such as [Kubernetes](https://kubernetes.io/), [Hugo](https://gohugo.io/), and [GitHub CLI](https://github.com/cli/cli) to name a few. [This list](https://github.com/spf13/cobra/blob/main/site/content/projects_using_cobra.md) contains a more extensive list of projects using Cobra.

Cobra is a library providing a simple interface to create powerful modern CLI interfaces similar to git & go tools.
```

But happy to switch `Framework` to `interface` even though thats semi redudent of the `I` in `CLI`

Would this work?

```
**[Cobra](https://github.com/spf13/cobra)** is the library we use to define our CLI. Commands are defined as Go structs with handlers, descriptions, and flags.
```

> - **[Cobra](https://github.com/spf13/cobra)** powers our CLI framework. Commands are defined as Go structs with handlers, descriptions, and flags.

Reviewer: > create powerful modern CLI interfaces
Yeah they do the redundancy in the part you quoted too :)

What you wrote would work 👍

> - **[Cobra](https://github.com/spf13/cobra)** powers our CLI framework. Commands are defined as Go structs with handlers, descriptions, and flags.

Author: ```suggestion
**[Cobra](https://github.com/spf13/cobra)** is the library we use to define our CLI. Commands are defined as Go structs with handlers, descriptions, and flags.
```

> - **[Viper](https://github.com/spf13/viper)** handles flag parsing, environment variables, and configuration files.

Author: Normally Cobra uses golang pFlag i think but Viper was built in tandem with Cobra to give more batteries included stuff

You can see our usage of it here - https://github.com/OpsLevel/cli/blob/main/src/cmd/root.go#L44

and then throughout the codebase as things like `viper.GetString("flag-name"`

> - **[Viper](https://github.com/spf13/viper)** handles flag parsing, environment variables, and configuration files.

Reviewer: OK, would be nice if that was clarified here in some way, otherwise their interaction is not super obvious.

> - **[Viper](https://github.com/spf13/viper)** handles flag parsing, environment variables, and configuration files.

Author: ```suggestion
- **[Viper](https://github.com/spf13/viper)** handles flag parsing, environment variables, and configuration files.  Examples of our usage of it are [here](https://github.com/OpsLevel/cli/blob/main/src/cmd/root.go#L44) and [here](https://github.com/OpsLevel/cli/blob/main/src/cmd/policy.go#L246)
```

- Modular command files live under `/cmd`, grouped by functionality (e.g., services, checks, etc.).
- Commands are registered to `rootCmd` via `init()` functions.
- 80% of our functionality is provided by opslevel-go and the purpose of this CLI is just to marshal data between the user and opslevel-go in a UX friendly way.
- Most commands follow the standard CRUD pattern `opslevel create ...`, `opslevel get ...`, `opslevel list ...`, `opslevel update ...`, `opslevel delete ...`, etc.

> - We have `opslevel beta` subcommand for experimental commands that are subject to removal.

Author: ```suggestion
- We have an `opslevel beta` subcommand for experimental commands that are subject to removal.
```

---

## Getting Started

### Clone the Repo

> If you're an external contributor, fork the repo:

Author: ```suggestion
If you're an external contributor fork the repo, then:
```

```bash
git clone https://github.com/YOUR_USERNAME/cli.git
cd cli
git checkout -b my-feature
```

If you're part of the OpsLevel org:

```bash
git clone git@github.com:OpsLevel/cli.git
cd cli
git checkout -b my-feature
```

> You might need to be given permissions to the repo, please reach out to team platform.

### Prerequisites

- [Task](https://taskfile.dev)
- An [OpsLevel API Token](https://app.opslevel.com/api_tokens)

You can use `task setup` to run an idempotent one-time setup.

Set your API token:

```sh
export OPSLEVEL_API_TOKEN=your_token_here
```

If you need to target a different environment, set the `OPSLEVEL_API_URL` environment variable:

```sh
export OPSLEVEL_API_URL=https://self-hosted.opslevel.dev/
```

---

## Development Workflow

### Run the CLI Locally

```sh

> go run main.go --help

Author: ```suggestion
cd ./src
go run main.go --help
```

```

This is the easiest way to test your changes live.

### Using a commit from an `opslevel-go` Branch

Sometimes you will need to make CLI changes in tandem with changes to [`opslevel-go`](https://github.com/OpsLevel/opslevel-go). 
To test changes from a local branch of [`opslevel-go`](https://github.com/OpsLevel/opslevel-go):

```sh
# Set up workspace using task
task workspace

# Switch to your feature branch in the submodule
git -C ./src/submodules/opslevel-go checkout --track origin/my-feature-branch
```

All CLI calls will now use your local `opslevel-go` code checked out into the submodule at `./src/submodules/opslevel-go`.
This way you can effectively work on both the CLI and `opslevel-go` in parallel if needed.

## Testing & Tooling

- Use `task test` to run tests locally
- Use `task lint` to check for code quality issues locally
- Use `task fix` to fix formatting, linting, go.mod, and update submodule all in one go

> Our CI pipeline will run `task ci` which can also be run locally to debug any issue that might only arrise in CI.

Author: ```suggestion
Our CI pipeline will run `task ci` which can also be run locally to debug any issue that might only arise in CI.
```

5. [Command Architecture & Style](#command-architecture--style)
6. [Testing & Tooling](#testing--tooling)
7. [Release Process](#release-process)

---

## About the CLI

The `opslevel-cli` is a command-line interface for interacting with the [OpsLevel](https://www.opslevel.com) API. It helps engineers automate, inspect, and manage their service catalog, ownership, checks, and more.

### Architecture

**[Cobra](https://github.com/spf13/cobra)** is the library we use to define our CLI. Commands are defined as Go structs with handlers, descriptions, and flags.
- **[Viper](https://github.com/spf13/viper)** handles flag parsing, environment variables, and configuration files.
- Modular command files live under `/cmd`, grouped by functionality (e.g., services, checks, etc.).
- Commands are registered to `rootCmd` via `init()` functions.
- 80% of our functionality is provided by opslevel-go and the purpose of this CLI is just to marshal data between the user and opslevel-go in a UX friendly way.
- Most commands follow the standard CRUD pattern `opslevel create ...`, `opslevel get ...`, `opslevel list ...`, `opslevel update ...`, `opslevel delete ...`, etc.

> - We have an `opslevel beta` subcommand for experimental commands that are subject to removal.

Reviewer: IMO s/removal/change

Author: ```suggestion
- We have an `opslevel beta` subcommand for experimental commands that are subject to change.
```
